File: src/java1/algorithms/binary/reverseBits/ReverseBits.java
// TC:O(32) ~ O(1) SC:O(1)
package java1.algorithms.binary.reverseBits;

public class ReverseBits {
    private static int reverseBits1(int n) {
        int reverseBits = 0;
        for(int i=0; i< 32; i++) {
            if(((n >> i) & 1) == 1) {
                reverseBits = reverseBits | (1 << (31 - i));
            }
        }
        return reverseBits;
    }

    private static int reverseBits2(int n) {
        int reverseBits = 0;
        for(int i=0; i< 32; i++) {
            reverseBits <<= 1;
            reverseBits |= (1 & n);
            n >>= 1;
        }
        return reverseBits;
    }

    public static void main(String[] args) {
        int num1 = 4;
        int num2 = 8;
        System.out.println(reverseBits1(num1));
        System.out.println(reverseBits2(num1));

        System.out.println(reverseBits1(num2));
        System.out.println(reverseBits2(num2));
    }
}

File: src/javascript/algorithms/binary/5.reverseBits/reverseBits.js
// TC:O(32) ~ O(1) SC:O(1)

function reverseBits1(num) {
    let reverseBitResult = 0;
    for(let i=0; i<32; i++) {
        let bit = (num >> i) & 1;
        if( bit === 1) {
            reverseBitResult = reverseBitResult | bit << 31-i;
        }
    }
    return reverseBitResult;
}

function reverseBits2(num) {
    let reverseBitResult = 0;
    for(let i=0; i< 32; i++) {
        reverseBitResult <<= 1;
        reverseBitResult |= (1 & num);
        num >>= 1;
    }
    return reverseBitResult;
}

let num1 = 4;
console.log(reverseBits1(num1));
console.log(reverseBits2(num1));

let num2 = 8;
console.log(reverseBits1(num2));
console.log(reverseBits2(num2));